aiokafka/producer.py

```diff
@@ -496,3 +496,39 @@ def _partition(self, topic, partition, key, value,
         available = list(self._metadata.available_partitions_for_topic(topic))
         return self._partitioner(
             serialized_key, all_partitions, available)
+
+    def create_batch(self, topic, partition=None):
+        """Create and return an empty MessageBatch.
+
+        The batch is not queued for send until submission to ``send_batch``.
+        Use of both ``create_batch`` and ``send`` together is discouraged as it
+        may lead to unexpected results.
+
+        Arguments:
+            topic (str): topic where the batch will be published
+            partition (int, optional): optionally specify a partition. If not
+                set, the partition will be selected using the configured
+                'partitioner' with key and value set to None.
+
+        Returns:
+            MessageBatch: empty batch to be filled and submitted by the caller.
+        """
+        partition = self._partition(topic, partition, None, None, None, None)
+        tp = TopicPartition(topic, partition)
+        return self._message_accumulator.batch_factory(tp)
```

Comment:
I don't think it's a good idea to use thus type for public Api. Lets just create a new, simple class, that will be used here. It can be an adapter around this one.

Reply:
Also we need docs for Batch class methods.

Reply:
What worries you about using this class for a public API?

The goal of this change is to reduce indirection and offer an interface "closer to the metal" specifically to support more flexible use-cases and, eventually, swapping parts for increased performance (e.g., one of my long term goals is to be able to swap out the Message class in use for one optimized to my use case). I'd rather fix what needs to be fixed in MessageBatch for it to work as a public interface than add another layer of indirection.

Reply:
Because that class is full of methods, that are not meant to be used, like `done`, `expired`, `wait_deliver`, `drain_ready` and so on. I want public API to be simple and not allow the user to shoot himself.

Reply:
Second, the `ttl` will be counted incorrectly if you use this class.

Reply:
Split MessageBatch into 2 classes: `ProducerBatch` and `MessageBatch`. The second is not aware of topic and partition, will only contain the resulting formatted buffer. The first will be aware of `tp`, `ttl`, futures.

Reply:
Okay.